src/GDALUtils.cpp

```diff
@@ -324,19 +415,151 @@ void BandReader::readBlock(int x, int y, uint8_t *data)
+    {
+        // Find the offset location in the source container.
+        int wholeRowElts = m_xTotalSize * ((y * m_yBlockSize) + row);
+        int partialRowElts = m_xBlockSize * x;
+
+        const uint8_t *srcStart = data +
+            ((wholeRowElts + partialRowElts) * m_eltSize);
+        const uint8_t *srcEnd = srcStart + (m_xBlockSize * m_eltSize);
+        std::copy(srcStart, srcEnd, dstStart);
+
+        // Blocks are always full-sized, even if only some of the data is valid,
+        // so we use m_xBlockSize instead of xWidth.
+        dstStart += (m_xBlockSize * m_eltSize);
+    }
+    if (m_band->WriteBlock(x, y, m_buf.data()) != CPLE_None)
+        throw CantWriteBlock();
+}
+
+Raster::Raster(const std::string& filename, const std::string& drivername)
+    : m_filename(filename)
+    , m_width(0)
+    , m_height(0)
+    , m_numBands(0)
+    , m_drivername(drivername)
+    , m_ds(0)
+{
+    m_forwardTransform.fill(0);
+    m_forwardTransform[1] = 1;
+    m_forwardTransform[5] = 1;
+    m_inverseTransform.fill(0);
+    m_inverseTransform[1] = 1;
+    m_inverseTransform[5] = 1;
+}
+
+
+Raster::Raster(const std::string& filename, const std::string& drivername,
+    const SpatialReference& srs, const std::array<double, 6> pixelToPos)
     : m_filename(filename)
-    , m_raster_x_size(0)
-    , m_raster_y_size(0)
-    , m_band_count(0)
+    , m_width(0)
+    , m_height(0)
+    , m_numBands(0)
+    , m_drivername(drivername)
+    , m_forwardTransform(pixelToPos)
+    , m_srs(srs)
     , m_ds(0)
+{}
+
+
+GDALError Raster::open(int width, int height, int numBands,
+    Dimension::Type type, double noData, StringList options)
 {
-    m_forward_transform.fill(0);
-    m_forward_transform[1] = 1;
-    m_forward_transform[5] = 1;
-    m_inverse_transform.fill(0);
-    m_inverse_transform[1] = 1;
-    m_inverse_transform[5] = 1;
+    if (m_drivername.empty())
+        m_drivername = "GTiff";
+
+    m_width = width;
+    m_height = height;
+    m_numBands = numBands;
+
+    if (!GDALInvGeoTransform(m_forwardTransform.data(),
+        m_inverseTransform.data()))
+    {
+        m_errorMsg = "Geotransform for raster '" + m_filename + "' not "
+           "invertible";
+        return GDALError::NotInvertible;
+    }
+
+    GDALDriver *driver = GetGDALDriverManager()->GetDriverByName(
+        m_drivername.data());
+    if (!driver)
+    {
+        m_errorMsg = "Driver '" + m_drivername + "' not found.";
+        return GDALError::DriverNotFound;
+    }
+
+    std::string item;
+    const char *itemp = driver->GetMetadataItem(GDAL_DCAP_CREATE);
+    if (itemp)
+        item = itemp;
+    if (item != "YES")
+    {
+        m_errorMsg = "Requested driver '" + m_drivername + "' does not "
+            "support file creation.";
+        return GDALError::InvalidDriver;
+    }
+
+    std::vector<const char *> opts;
+
+    for (size_t i = 0; i < options.size(); ++i)
+    {
+        if (options[i].find("INTERLEAVE") == 0)
```

Comment:
Need a note about this in the documentation. Is the reason this doesn't work is you're copying directly into the array and expect a certain arrangement?